### PR TITLE
Require translogUUID when reading global checkpoint

### LINKED ISSUE #28435: [CI] CorruptedFileIT#testReplicaCorruption fails on Windows
CI: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-windows-compatibility/1179/console
Log: [testReplicaCorruption.txt](https://github.com/elastic/elasticsearch/files/1676317/testReplicaCorruption.txt)

We intentionally corrupted every segment files in `#testReplicaCorruption`. This caused file-based recovery occurring as replicas could not to read commit snapshot. However, there was one replica `[test][7]` failed to clean up its store after index files were copied. The corruption marker `corrupted_WzIGIIeIQvqVgKvEIVtRwA` was deleted but still in the listing and inaccessible(on Windows).

```
1> Caused by: java.nio.file.AccessDeniedException: C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\server\build\testrun\integTest\J1\temp\org.elasticsearch.index.store.CorruptedFileIT_883F8FC2C6E448B5-001\tempDir-002\data\nodes\0\indices\0tscMvZ-Q-21DRHviJsJxw\7\index\corrupted_WzIGIIeIQvqVgKvEIVtRwA
  1> 	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:83) ~[?:?]
  1> 	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97) ~[?:?]
  1> 	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102) ~[?:?]
  1> 	at sun.nio.fs.WindowsFileSystemProvider.newByteChannel(WindowsFileSystemProvider.java:230) ~[?:?]
  1> 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newByteChannel(FilterFileSystemProvider.java:212) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newByteChannel(FilterFileSystemProvider.java:212) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newByteChannel(FilterFileSystemProvider.java:212) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.mockfile.HandleTrackingFS.newByteChannel(HandleTrackingFS.java:240) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newByteChannel(FilterFileSystemProvider.java:212) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.mockfile.HandleTrackingFS.newByteChannel(HandleTrackingFS.java:240) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at java.nio.file.Files.newByteChannel(Files.java:361) ~[?:1.8.0_144]
  1> 	at java.nio.file.Files.newByteChannel(Files.java:407) ~[?:1.8.0_144]
  1> 	at org.apache.lucene.store.SimpleFSDirectory.openInput(SimpleFSDirectory.java:77) ~[lucene-core-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:43]
  1> 	at org.apache.lucene.util.LuceneTestCase.slowFileExists(LuceneTestCase.java:2709) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.store.MockDirectoryWrapper.openInput(MockDirectoryWrapper.java:749) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:43]
  1> 	at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:99) ~[lucene-core-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:43]
  1> 	at org.apache.lucene.store.Directory.openChecksumInput(Directory.java:119) ~[lucene-core-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:43]
  1> 	at org.elasticsearch.index.store.Store.failIfCorrupted(Store.java:615) ~[main/:?]
  1> 	at org.elasticsearch.index.store.Store.failIfCorrupted(Store.java:607) ~[main/:?]
  1> 	at org.elasticsearch.index.store.Store.getMetadata(Store.java:281) ~[main/:?]
  1> 	at org.elasticsearch.index.store.Store.getMetadata(Store.java:253) ~[main/:?]
  1> 	at org.elasticsearch.index.store.Store.cleanupAndVerify(Store.java:683) ~[main/:?]
  1> 	at org.elasticsearch.indices.recovery.RecoveryTarget.cleanFiles(RecoveryTarget.java:444) [main/:?]
  1> 	... 10 more
```

As the target failed to clean up, it retried another recovery. At that moment the corrupted marker was removed and the copied commit is safe, hence the replica starts sequence-based recovery. Unfortunately, translog and index commit on the target were mismatch because the Lucene commit was copied from the primary. This replica `[test][7]` could not recover from primary.

```
  1> Caused by: org.elasticsearch.index.translog.TranslogCorruptedException: expected shard UUID [32 5f 7a 61 42 43 56 4d 53 4f 47 4a 6a 35 30 4f 6c 69 65 43 44 51] but got: [54 62 51 6c 30 6a 49 4f 53 50 53 68 5a 76 58 2d 2d 44 46 70 58 41] this translog file belongs to a different translog. path:C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\server\build\testrun\integTest\J1\temp\org.elasticsearch.index.store.CorruptedFileIT_883F8FC2C6E448B5-001\tempDir-002\data\nodes\0\indices\0tscMvZ-Q-21DRHviJsJxw\7\translog\translog-2.tlog
  1> 	at org.elasticsearch.index.translog.TranslogReader.open(TranslogReader.java:131) ~[main/:?]
  1> 	at org.elasticsearch.index.translog.Translog.openReader(Translog.java:292) ~[main/:?]
  1> 	at org.elasticsearch.index.translog.Translog.recoverFromFiles(Translog.java:238) ~[main/:?]
  1> 	at org.elasticsearch.index.translog.Translog.<init>(Translog.java:176) ~[main/:?]
  1> 	at org.elasticsearch.index.engine.InternalEngine.openTranslog(InternalEngine.java:491) ~[main/:?]
  1> 	at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:180) ~[main/:?]
  1> 	at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:151) ~[main/:?]
  1> 	at org.elasticsearch.index.engine.InternalEngineFactory.newReadWriteEngine(InternalEngineFactory.java:25) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.newEngine(IndexShard.java:2149) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.createNewEngine(IndexShard.java:2131) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.innerOpenEngineAndTranslog(IndexShard.java:1361) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.openIndexAndSkipTranslogRecovery(IndexShard.java:1327) ~[main/:?]
  1> 	at org.elasticsearch.indices.recovery.RecoveryTarget.prepareForTranslogOperations(RecoveryTarget.java:371) ~[main/:?]
  1> 	at org.elasticsearch.indices.recovery.PeerRecoveryTargetService$PrepareForTranslogOperationsRequestHandler.messageReceived(PeerRecoveryTargetService.java:396) ~[main/:?]
  1> 	at org.elasticsearch.indices.recovery.PeerRecoveryTargetService$PrepareForTranslogOperationsRequestHandler.messageReceived(PeerRecoveryTargetService.java:390) ~[main/:?]
```

### PR DESCRIPTION
**Pull request by dnhatn**
Today we use the persisted global checkpoint to calculate the starting seqno in peer-recovery. However we do not check whether the translog actually belongs to the existing Lucene index when reading the global checkpoint. In some rare cases if the translog does not match the Lucene index, that recovering replica won't be able to complete its recovery. This can happen as follows.

1. A replica executes a file-based recovery
2. Index files are copied to replica but it crashed before finishing the recovery
3. Replica starts recovery again with a sequence-based recovery as the copied commit is safe
4. Replica fails to open engine because its translog and Lucene index are not matched
5. Replica won't be able to recover from primary

This commit enforces the translogUUID requirement when reading the global checkpoint directly from the checkpoint file.

Relates #28435

### CODE REVIEW

**Comment by bleskes** [comment]
@dnhatn I was thinking to add a translog uuid parameter to org.elasticsearch.index.translog.Translog#readGlobalCheckpoint , so you always have to indicate which translog you expect when you just go and read the global checkpoint. It requires some juggling in IndexShard (where we read it before opening the engine) and I'm not sure how that will look. I wonder whether we should put our time in the store corruption marker reflecting a broken translog. Let's discuss.

**Comment by dnhatn** [comment]
Closing this in favour of #28604

**Comment by dnhatn** [comment]
@bleskes 
Yannick and I discussed this and agreed to fix the test by #28604. I will think more about the translog corruption issue.

**Comment by dnhatn** [comment]
@bleskes I've added translogUUID requirement when reading the global checkpoint. Can you please have a look? Thank you!

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 1]
```diff
@@ -1682,19 +1686,33 @@ public Exception getTragicException() {
...
+     * @throws IOException                if an I/O exception occurred reading the checkpoint
+     * @throws TranslogCorruptedException if the translog is corrupted or mismatched with the given uuid
      */
-    public static final long readGlobalCheckpoint(final Path location) throws IOException {
-        return readCheckpoint(location).globalCheckpoint;
+    public static long readGlobalCheckpoint(final Path location, final String expectedTranslogUUID) throws IOException {
+        final Checkpoint checkpoint = readCheckpoint(location);
+        // We open files in reverse order in order to validate translog uuid before we start traversing the translog based on
+        // the generation id we found in the lucene commit. This gives for better error messages if the wrong translog was found.
+        for (long i = checkpoint.generation - 1; i >= checkpoint.minTranslogGeneration; i--) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java]: I think we can just read the uuid of the generation associated with the checkpoint? I think this is overly fanatic, no? (I want to make a more complete validation at one place in a different PR - complete in the sense, check multiple lucene commits and multiple generations.

**Review by bleskes — CHANGES_REQUESTED** [review]
I left a very minor comment.

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 2]
```diff
@@ -1682,19 +1686,30 @@ public Exception getTragicException() {
...
-        return readCheckpoint(location).globalCheckpoint;
+    public static long readGlobalCheckpoint(final Path location, final String expectedTranslogUUID) throws IOException {
+        final Checkpoint checkpoint = readCheckpoint(location);
+        // We need to open at least translog reader to validate the translogUUID.
+        final Path translogFile = location.resolve(getFilename(checkpoint.generation));
+        if (Files.exists(translogFile) == false) {
+            throw new TranslogCorruptedException("Translog file [" + translogFile + "] doesn't exist; checkpoint [" + checkpoint + "]");
+        }
+        try (TranslogReader reader = openReader(translogFile,
+            Checkpoint.read(location.resolve(getCommitCheckpointFileName(checkpoint.generation))), expectedTranslogUUID)) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java]: why do we need to read the checkpoint again? I think we need to be consistent here - either with throw the underlying exception (and thus not protect about FileNotFoundException) or catch everything and rethrow a corruption exception. I tend toward the later..

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 3]
```diff
@@ -1703,11 +1703,12 @@ public static long readGlobalCheckpoint(final Path location, final String expect
         final Checkpoint checkpoint = readCheckpoint(location);
         // We need to open at least translog reader to validate the translogUUID.
         final Path translogFile = location.resolve(getFilename(checkpoint.generation));
-        if (Files.exists(translogFile) == false) {
-            throw new TranslogCorruptedException("Translog file [" + translogFile + "] doesn't exist; checkpoint [" + checkpoint + "]");
-        }
-        try (TranslogReader reader = openReader(translogFile,
-            Checkpoint.read(location.resolve(getCommitCheckpointFileName(checkpoint.generation))), expectedTranslogUUID)) {
+        try (TranslogReader reader = openReader(translogFile, checkpoint, expectedTranslogUUID)) {
+
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java]: add a comment about what we're doing here?

**Review by bleskes — APPROVED** [review]

**Comment by dnhatn** [comment]
Thanks @bleskes for your review.

**Pull request merged**